### PR TITLE
[SPARK-23744][CORE]Fix memory leak in ReadableChannelFileRegion

### PR DESCRIPTION
**Pull request by 10110346**
## What changes were proposed in this pull request?
In the class `ReadableChannelFileRegion`,  the `buffer` is direct memory, we should  modify `deallocate` to free it, and `deallocate`  will be called by `release`

## How was this patch tested?
existing unit test

### CODE REVIEW

**Comment by 10110346** [comment]
retest this please

**Comment by 10110346** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 1]
```diff
@@ -301,7 +301,10 @@ private class ReadableChannelFileRegion(source: ReadableByteChannel, blockSize:
     written
   }
 
-  override def deallocate(): Unit = source.close()
+  override def deallocate() {
+    source.close()
+    StorageUtils.dispose(buffer)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: IIRC direct byte buffer can be GCed, cc @kiszk

File: core/src/test/scala/org/apache/spark/storage/DiskStoreSuite.scala [hunk 2]
```diff
@@ -197,7 +197,7 @@ class DiskStoreSuite extends SparkFunSuite {
     while (region.transfered() < region.count()) {
       region.transferTo(byteChannel, region.transfered())
     }
-
+    region.release()
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/storage/DiskStoreSuite.scala]: why this change?

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala (hunk 1 above)
Comment (10110346) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: In fact, direct memory is also off-heap memory, 
if we config this parameter "-XX:+DisableExplicitGC",the memory will not be reclaimed

File: core/src/test/scala/org/apache/spark/storage/DiskStoreSuite.scala (hunk 2 above)
Comment (10110346) [reply core/src/test/scala/org/apache/spark/storage/DiskStoreSuite.scala]: I think we should release the direct memory in unit test

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala (hunk 1 above)
Comment (kiszk) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: Sorry for being late since I overlooked.

While `ByteBuffer.allocateDirect` allocates memory in off-heap memory, the memory will be also reclaimed when the ByteBuffer object will be reclaimed.
Regarding `-XX:+DisableExplicitGC`, as described [here](http://www.oracle.com/technetwork/java/javase/tech/vmoptions-jsp-140102.html), this option just disables `System.gc()`. Other GC operations may reclaim the memory in off-heap allocated by `ByteBuffer.allocateDirect`.
>By default calls to System.gc() are enabled (-XX:-DisableExplicitGC). Use -XX:+DisableExplicitGC to disable calls to System.gc(). Note that the JVM still performs garbage collection when necessary.

Comment (kiszk) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: In other words, I think that the off-heap in `buffer` will not be leaked if `buffer` will be reclaimed.

Comment (10110346) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: I see,thanks
